core/templates/pages/exploration-player-page/services/stats-reporting.service.ts

```diff
@@ -16,344 +16,401 @@
+import { PlaythroughService } from 'services/playthrough.service';
+import { SiteAnalyticsService } from 'services/site-analytics.service';
+import { Stopwatch, StopwatchObjectFactory } from
+  'domain/utilities/StopwatchObjectFactory';
+import { UrlInterpolationService } from
+  'domain/utilities/url-interpolation.service';
 
-    var createDefaultStateStatsMapping = function(stateName) {
-      aggregatedStats.state_stats_mapping[stateName] = {
-        total_answers_count: 0,
-        useful_feedback_count: 0,
-        total_hit_count: 0,
-        first_hit_count: 0,
-        num_times_solution_viewed: 0,
-        num_completions: 0
-      };
-    };
+interface IStateStats {
+  /* eslint-disable camelcase */
+  total_answers_count: number;
+  useful_feedback_count: number;
+  total_hit_count: number;
+  first_hit_count: number;
+  num_times_solution_viewed: number;
+  num_completions: number;
+  /* eslint-enable camelcase */
+}
+interface IAggregatedStats {
+  /* eslint-disable camelcase */
+  num_starts: number;
+  num_completions: number;
+  num_actual_starts: number;
+  state_stats_mapping: {[stateName: string]: IStateStats};
+  /* eslint-enable camelcase */
+}
 
-    var getFullStatsUrl = function(urlIdentifier) {
-      try {
-        return UrlInterpolationService.interpolateUrl(
-          STATS_REPORTING_URLS[urlIdentifier], {
-            exploration_id: explorationId
-          });
-      } catch (e) {
-        var additionalInfo = ('\nUndefined exploration id error debug logs:' +
-          '\nThe event being recorded: ' + urlIdentifier +
-          '\nExploration ID: ' + ContextService.getExplorationId()
-        );
-        if (currentStateName) {
-          additionalInfo += ('\nCurrent State name: ' + currentStateName);
-        }
-        if (nextExpId) {
-          additionalInfo += ('\nRefresher exp id: ' + nextExpId);
-        }
-        if (previousStateName && nextStateName) {
-          additionalInfo += ('\nOld State name: ' + previousStateName +
-            '\nNew State name: ' + nextStateName);
-        }
-        e.message += additionalInfo;
-        throw e;
-      }
+@Injectable({
+  providedIn: 'root'
+})
+export class StatsReportingService {
+  constructor(
+      private contextService: ContextService,
+      private http: HttpClient,
+      private messengerService: MessengerService,
+      private playthroughService: PlaythroughService,
+      private siteAnalyticsService: SiteAnalyticsService,
+      private stopwatchObjectFactory: StopwatchObjectFactory,
+      private urlInterpolationService: UrlInterpolationService) {
+    this.editorPreviewMode = (
+      this.contextService.isInExplorationEditorPage());
+    this.questionPlayerMode = this.contextService.isInQuestionPlayerMode();
+    this.refreshAggregatedStats();
+  }
+
+  explorationId: string = null;
+  explorationTitle: string = null;
+  explorationVersion: number = null;
+  sessionId: string = null;
```

Comment:
Ditto here and all the class variables that control some logic in code or are returned as data to another directive/service.

Reply:
Done.